reconcile/openshift_serviceaccount_tokens.py

```diff
@@ -47,117 +53,147 @@ def get_tokens_for_service_account(
+def write_outputs_to_vault(
+    vault_client: VaultClient, vault_path: str, ri: ResourceInventory
+) -> None:
     integration_name = QONTRACT_INTEGRATION.replace("_", "-")
-    vault_client = VaultClient()
     for cluster, namespace, _, data in ri:
         for name, d_item in data["desired"].items():
             body_data = d_item.body["data"]
             # write secret to per-namespace location
             secret_path = (
-                f"{vault_path}/{integration_name}/" + f"{cluster}/{namespace}/{name}"
+                f"{vault_path}/{integration_name}/{cluster}/{namespace}/{name}"
             )
             secret = {"path": secret_path, "data": body_data}
-            vault_client.write(secret)
+            vault_client.write(secret)  # type: ignore
             # write secret to shared-resources location
-            secret_path = (
-                f"{vault_path}/{integration_name}/" + f"shared-resources/{name}"
-            )
+            secret_path = f"{vault_path}/{integration_name}/shared-resources/{name}"
             secret = {"path": secret_path, "data": body_data}
-            vault_client.write(secret)
+            vault_client.write(secret)  # type: ignore
 
 
-def canonicalize_namespaces(namespaces):
+def canonicalize_namespaces(namespaces: Iterable[NamespaceV1]) -> list[NamespaceV1]:
     canonicalized_namespaces = []
-    for namespace_info in namespaces:
-        if ob.is_namespace_deleted(namespace_info):
-            continue
-        ob.aggregate_shared_resources(namespace_info, "openshiftServiceAccountTokens")
-        openshift_serviceaccount_tokens = namespace_info.get(
-            "openshiftServiceAccountTokens"
-        )
-        if openshift_serviceaccount_tokens:
-            canonicalized_namespaces.append(namespace_info)
-            for sat in openshift_serviceaccount_tokens:
-                canonicalized_namespaces.append(sat["namespace"])
+    for namespace in namespaces:
+        ob.aggregate_shared_resources_typed(namespace)
+        if namespace.openshift_service_account_tokens:
+            canonicalized_namespaces.append(namespace)
+            for sat in namespace.openshift_service_account_tokens:
+                if f"{sat.namespace.cluster.name}/{sat.namespace.name}" in [
+                    f"{ns.cluster.name}/{ns.name}" for ns in canonicalized_namespaces
+                ]:
+                    # do not add duplicate namespaces to speed up fetch_current_state
+                    continue
+                canonicalized_namespaces.append(
+                    # convert openshiftServiceAccountTokens.namespace to NamespaceV1
+                    NamespaceV1(
+                        **sat.namespace.dict(by_alias=True),
+                        sharedResources=None,
+                        openshiftServiceAccountTokens=None,
+                    )
+                )
 
     return canonicalized_namespaces
 
 
+def get_serviceaccount_tokens(query_func: Callable) -> list[NamespaceV1]:
```

Comment:
naming is a little confusing, as return is a list of namespace instead of tokens

Reply:
Applied in f5c17c5